### PR TITLE
Upgradeable contracts

### PR DESCRIPTION
**Pull request by fbielejec**
Demonstrates upgrading ink! contracts with a set_code

### CODE REVIEW

**Review by JoshOrndorff — APPROVED** [review]
I've taken a quick look and I get the general idea. I definitely support teaching the concept of upgradeable contracts. I actually didn't know pallet contracts allowed that. I'm excited that students will first encounter this at the smart contract level so they are better prepared to understand runtime upgrades which are really the same thing, but sometimes feel harder to understand or more "mind bending".

Overall, I'm fine with it being merged, but all this ink content still feels out of place to me in this repo that is mostly about a PoW node.

I understand that you want to test this stuff against the node, so, again I'm fine to have it all here especially for the upcoming wave. But ultimately I wonder if the ink content should have a separate loosely-coupled home.

My plan for the contract writing workshop (at least for now) is to take some of the contract ideas covered here and reduce them to starter code for students to implement on their own.

**Comment by fbielejec** [comment]
> I've taken a quick look and I get the general idea. I definitely support teaching the concept of upgradeable contracts. I actually didn't know pallet contracts allowed that. I'm excited that students will first encounter this at the smart contract level so they are better prepared to understand runtime upgrades which are really the same thing, but sometimes feel harder to understand or more "mind bending".
> 
> Overall, I'm fine with it being merged, but all this ink content still feels out of place to me in this repo that is mostly about a PoW node.
> 
> I understand that you want to test this stuff against the node, so, again I'm fine to have it all here especially for the upcoming wave. But ultimately I wonder if the ink content should have a separate loosely-coupled home.
> 
> My plan for the contract writing workshop (at least for now) is to take some of the contract ideas covered here and reduce them to starter code for students to implement on their own.

Exactly, the idea is to have it tested against the node, and then create a submodule in the contracts repository. 
None of it is production quality code, far from, and there is plenty of improvements and extensions possible - happy to list them

File: contracts/new_a/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "a"
+version = "0.2.0"
+authors = ["fbielejec"]
+edition = "2021"
+
+[dependencies]
+ink = { version = "=4.2.1",  default-features = false }
+# old_a = { package = "a", path = "../old_a", default-features = false, features = ["ink-as-dependency"] }
```
Comment (pmikolajczyk41) [inline contracts/new_a/Cargo.toml:9]: leftover?

File: contracts/new_a/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "a"
```
Comment (pmikolajczyk41) [inline contracts/new_a/Cargo.toml:2]: I would be a bit more verbose xD

File: contracts/old_a/lib.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            data.field_1 = field_1;
+            data.field_2 = field_2;
+            self.old_state.set(&data);
+            Ok(())
+        }
+
+        /// Upgrades contract code
+        #[ink(message)]
+        pub fn set_code(&mut self, code_hash: [u8; 32], callback: Option<Selector>) -> Result<()> {
+            set_code_hash(&code_hash)?;
```
Comment (pmikolajczyk41) [inline contracts/old_a/lib.rs:69]: do we want any access control here?

File: contracts/new_a/lib.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                new_state.field_3,
+            )
+        }
+
+        /// Performs a contract storage migration.
+        ///
+        /// Call it only once
+        #[ink(message, selector = 0x4D475254)]
+        pub fn migrate(&mut self) -> Result<()> {
+            if let Some(_old_state @ OldState { field_1, field_2 }) = get_contract_storage(&123)? {
```
Comment (pmikolajczyk41) [inline contracts/new_a/lib.rs:77]: `_old_state @` not necessary, right?

**Review by pmikolajczyk41 — APPROVED** [review]

File: contracts/old_a/lib.rs (hunk 3 above)
Comment (fbielejec) [reply contracts/old_a/lib.rs:69]: its just a demo, did not want to complicate it with code that does not serve the purpose

File: contracts/new_a/Cargo.toml (hunk 1 above)
Comment (fbielejec) [reply contracts/new_a/Cargo.toml:9]: removed

File: contracts/new_a/lib.rs (hunk 4 above)
Comment (fbielejec) [reply contracts/new_a/lib.rs:77]: no, removed

**Pull request merged**